command/action.go

```diff
@@ -0,0 +1,412 @@
+		return 1
+	}
+
+	if task != "" {
+		err = validateTaskExistsInAllocation(task, alloc)
+	} else {
+		task, err = lookupAllocTask(alloc)
+	}
+	if err != nil {
+		l.Ui.Error(err.Error())
+		return 1
+	}
+	l.Ui.Output(fmt.Sprintf("Task set: %s", task))
+
+	if !stdinOpt {
+		l.Stdin = bytes.NewReader(nil)
+	}
+
+	if l.Stdin == nil {
+		l.Stdin = os.Stdin
+	}
+
+	if l.Stdout == nil {
+		l.Stdout = os.Stdout
+	}
+
+	if l.Stderr == nil {
+		l.Stderr = os.Stderr
+	}
+
+	action := args[0]
+	// Convert action to command here.
+	// TODO: Noting that this is also being (redundantly) done in client/alloc_endpoint.go's execImpl.
+	// I could pass the action name all the way down, but that would be maybe polluting the param space of
+	// api/allocations.go, api/allocations_exec.go, etc., since alloc exec normally would never use an Action
+	// except by via this command.
+	// TODO: update: I think I'll try that anyway.
+
+	code, err := l.execImpl(client, alloc, task, job, action, ttyOpt, escapeChar, l.Stdin, l.Stdout, l.Stderr)
+	if err != nil {
+		l.Ui.Error(fmt.Sprintf("failed to exec into task: %v", err))
+		return 1
+	}
+
+	return code
+}
+
+// execImpl invokes the Alloc Exec api call, it also prepares and restores terminal states as necessary.
+func (l *ActionCommand) execImpl(client *api.Client, alloc *api.Allocation, task string, job string, action string, tty bool,
+	escapeChar string, stdin io.Reader, stdout, stderr io.WriteCloser) (int, error) {
+
+	l.Ui.Output(fmt.Sprintf("Impl command: %s", l.Name()))
+
+	sizeCh := make(chan api.TerminalSize, 1)
+
+	ctx, cancelFn := context.WithCancel(context.Background())
+	defer cancelFn()
+
+	// When tty, ensures we capture all user input and monitor terminal resizes.
+	if tty {
+		if stdin == nil {
+			return -1, fmt.Errorf("stdin is null")
+		}
+
+		inCleanup, err := setRawTerminal(stdin)
+		if err != nil {
+			return -1, err
+		}
+		defer inCleanup()
+
+		outCleanup, err := setRawTerminalOutput(stdout)
+		if err != nil {
+			return -1, err
+		}
+		defer outCleanup()
+
+		sizeCleanup, err := watchTerminalSize(stdout, sizeCh)
+		if err != nil {
+			return -1, err
+		}
+		defer sizeCleanup()
+
+		if escapeChar != "" {
+			stdin = escapingio.NewReader(stdin, escapeChar[0], func(c byte) bool {
+				switch c {
+				case '.':
+					// need to restore tty state so error reporting here
+					// gets emitted at beginning of line
+					outCleanup()
+					inCleanup()
+
+					stderr.Write([]byte("\nConnection closed\n"))
+					cancelFn()
+					return true
+				default:
+					return false
+				}
+			})
+		}
+	}
+
+	// TODO: establish command here from action.
+
+	signalCh := make(chan os.Signal, 1)
+	signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		for range signalCh {
+			cancelFn()
+		}
+	}()
+
+	// TODO: is make([]string, 0) the right thing to pass an empty command here?
+	return client.Allocations().Exec(ctx,
+		alloc, task, tty, make([]string, 0), action, stdin, stdout, stderr, sizeCh, nil)
```

Comment:
This should be going to a new `client.Job().Action(...)` that calls the `command/agent/job_endpoint.go` for the actions HTTP API.